homeassistant/components/samsungtv/bridge.py

```diff
@@ -0,0 +1,249 @@
+"""samsungctl and samsungtvws bridge classes."""
+from abc import ABC, abstractmethod
+
+import requests
+from requests import RequestException
+from samsungctl import Remote
+from samsungctl.exceptions import AccessDenied, ConnectionClosed, UnhandledResponse
+from samsungtvws import SamsungTVWS
+from websocket import WebSocketException
+
+from .const import (
+    LOGGER,
+    RESULT_AUTH_MISSING,
+    RESULT_NOT_SUCCESSFUL,
+    RESULT_NOT_SUPPORTED,
+    RESULT_SUCCESS,
+)
+
+
+class SamsungTVBridge(ABC):
+    """The Base Bridge abstract class."""
+
+    @staticmethod
+    def get_bridge(config):
+        """Get Bridge instance."""
+        if config:
+            if config["method"] == "legacy":
+                return SamsungTVLegacyBridge(config)
+            if config["method"] == "websocket":
+                return SamsungTVWSBridge(config)
+        return None
+
+    def __init__(self, config):
+        """Initialize Bridge."""
+        self.port = None
+        self.token = None
+        self.config = config
+        self.method = config["method"]
+        self.host = config["host"]
+        self._remote = None
+        self._callback = None
+
+    def register_reauth_callback(self, func):
+        """Register a callback function."""
+        self._callback = func
+
+    @abstractmethod
+    def try_connect(self, port):
+        """Try to connect to the TV."""
+
+    @abstractmethod
+    def is_on(self):
+        """Tells if the TV is on."""
+
+    def send_key(self, key):
+        """Send a key to the tv and handles exceptions."""
+        try:
+            # recreate connection if connection was dead
+            retry_count = 1
+            for _ in range(retry_count + 1):
+                try:
+                    self._send_key(key)
+                    break
+                except (
+                    ConnectionClosed,
+                    BrokenPipeError,
+                    WebSocketException,
+                ):
+                    # BrokenPipe can occur when the commands is sent to fast
+                    # WebSocketException can occur when timed out
+                    self._remote = None
+        except (UnhandledResponse, AccessDenied):
+            # We got a response so it's on.
+            LOGGER.debug("Failed sending command %s", key, exc_info=True)
+        except OSError:
+            # Different reasons, e.g. hostname not resolveable
+            pass
+
+    @abstractmethod
+    def _send_key(self, key):
+        """Send the key."""
+
+    @abstractmethod
+    def _get_remote(self):
+        """Get Remote object."""
+
+    def close_remote(self):
+        """Close remote object."""
+        self._get_remote().close()
+
+    def _notify_callback(self):
```

Comment:
The notify callback is only called from the legacy bridge.

Reply:
Don't know yet when to trigger it

Reply:
Added even for websocket bridge.